### ISSUE TITLE
Warning: DOMXPath::query(): Recursion limit exceeded

### ISSUE THREAD

**Issue by matteorossi-thespacesm** [magento/magento2#39273]
### Preconditions and environment

Magento 2.4.5-p10

### Steps to reproduce

Import a large number of products, e.g. 2000 products

### Expected result

Import succeds without warnings

### Actual result

Exception: Warning: DOMXPath::query(): **Recursion limit exceeded** in vendor/magento/framework/Validator/HTML/ConfigurableWYSIWYGValidator.php on line 125

### Additional information

Magento\Framework\Validator\HTML\ConfigurableWYSIWYGValidator

method validateConfigured(\DOMXPath $xpath) is changed as:

//Validating tags
**$this->allowedTags = array_merge($this->allowedTags, ["body", "html"]);**
$found = $xpath->query(
    '//*['
        . implode(
            ' and ',
            array_map(
                function (string $tag): string {
                    return "name() != '$tag'";
                },
                $this->allowedTags
            )
        )
        .']'
);

$this->allowedTags is  charged with two items "body" and "html" every time validateConfigured is called, so in the end xpath recursion limit is exceeded and such warning appears.
Maybe such bug is not triggered in test because recursion limit is quite high.
It is **MANDATORY** to revert method to the previous version:

//Validating tags
$this->allowedTags = array_merge($this->allowedTags, ["body", "html"]);
$found = $xpath->query(
    '//*['
        . implode(
            ' and ',
            array_map(
                function (string $tag): string {
                    return "name() != '$tag'";
                },
                $this->allowedTags
            )
        )
        .']'
);

### Triage and priority

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Bravo** [comment]
Hi @matteorossi-thespacesm,

Thanks for your reporting and collaboration.

We have verified the issue in Latest 2.4-develop instance and the issue is not reproducible.Kindly refer the attached video.



https://github.com/user-attachments/assets/6c9a61e3-d5c3-4a67-af4b-c9eab3c9991c




Import was done successfully without warnings.

Kindly recheck the issue in Latest 2.4-develop instance and elaborate the steps to reproduce if the issue is still reproducible.

Thanks.

**Comment by engcom-Bravo** [comment]
Hi @matteorossi-thespacesm,

This issue is being closed since it has not been updated in a long time.Please feel free to reopen or raise a new ticket if the issue still exists.

Thanks.

**Comment by dsengsqli** [comment]
I just updated to 2.4-5-p10 and also have the issue

**Comment by matteorossi-thespacesm** [comment]
Update: the issue is present on every patched version, including magento 2.4.7-p3.

Such issue is caused by a code fragment as reported above.
If anyone is logging $this->allowedTags field content, one will see such field growing due to duplicated "body" and "html" array items, at the point that recursion limit exceeded appears if one has xdebug extension installed and configured.
Otherwise, it is possibile to experience SEGFAULT errors.

**Comment by sergiirepin** [comment]
Can confirm the issue on  Adobe Commerce ver. 2.4.6-p8.

@engcom-Bravo 
To reproduce try to import +5k products.

**Comment by daniel-donselaar-pon** [comment]
The issue is this piece of code:
`$this->allowedTags = array_merge($this->allowedTags, ["body", "html"]);`

Located in `vendor/magento/framework/Validator/HTML/ConfigurableWYSIWYGValidator.php:113`

It keeps putting the 'body' and 'html' in the array, which becomes very big.

Replace it with this:

```
if (!in_array('body', $this->allowedTags)) {
    $this->allowedTags ['body'] = 'body';
}

if (!in_array('html', $this->allowedTags)) {
    $this->allowedTags ['html'] = 'html';
}
```

That should solve the problem.

**Comment by matteorossi-thespacesm** [comment]
> The issue is this piece of code: `$this->allowedTags = array_merge($this->allowedTags, ["body", "html"]);`
> 
> Located in `vendor/magento/framework/Validator/HTML/ConfigurableWYSIWYGValidator.php:113`
> 
> It keeps putting the 'body' and 'html' in the array, which becomes very big.
> 
> Replace it with this:
> 
> ```
> if (!in_array('body', $this->allowedTags)) {
>     $this->allowedTags ['body'] = 'body';
> }
> 
> if (!in_array('html', $this->allowedTags)) {
>     $this->allowedTags ['html'] = 'html';
> }
> ```
> 
> That should solve the problem.

Adding a conditional statement would make performance worse to me.

**Comment by matteorossi-thespacesm** [comment]
> I just can't understand why the approach prior to 2.4.7-p3 was changed. Can't we just revert the `ConfigurableWYSIWYGValidator` back to the following:
> 
> ```
>     private function validateConfigured(\DOMXPath $xpath): void
>     {
>         //Validating tags
>         $found = $xpath->query(
>             '//*['
>                 . implode(
>                     ' and ',
>                     array_map(
>                         function (string $tag): string {
>                             return "name() != '$tag'";
>                         },
>                         array_merge($this->allowedTags, ['body', 'html'])
>                     )
>                 )
>                 .']'
>         );
> ...
> ```
> 
> It seems to me that `array_merge($this->allowedTags, ['body', 'html'])` handles this properly, doesn't it?

It does. IMHO method code was modified to improve performance, deleting array_merge call.
Unfortunately it has such a side effect, that allowedTags array explodes.

**Comment by magnogaspar** [comment]
> ConfigurableWYSIWYGValidator

I see, but if the point is performance, I think adding 'body' and 'html' to the class constructor would be a better approach.

Something like:

```php
    public function __construct(
        array $allowedTags,
        array $allowedAttributes = [],
        array $attributesAllowedByTags = [],
        array $attributeValidators = [],
        array $tagValidators = []
    ) {
        if (empty(array_filter($allowedTags))) {
            throw new \InvalidArgumentException('List of allowed HTML tags cannot be empty');
        }
        $this->allowedTags = array_unique(array_merge($allowedTags, ['body', 'html']));
        $this->allowedAttributes = array_unique($allowedAttributes);
...
```

Anyway: I'm facing this issue and I had to use a patch to fix it. It needs to be reopened.

**Comment by hostep** [comment]
@engcom-Bravo, @nathanjosiah, @ihor-sviziev: can we get this issue re-openend? A colleague of mine ran into this problem as well on Magento 2.4.7-p3

It seems like this commit (https://github.com/magento/magento2/commit/42577bc5ab88ce74706d16356825440c9cdce9ba) causes the issue, which was introduced in the latest security patches.

I don't have steps to reproduce yet, if somebody in this thread has clear steps to reproduce and would like to post them, that would be appreciated :)

**Comment by nathanjosiah** [comment]
This is fixed within the upcoming February release versions in scope of internal ticket AC-13493 and related backports.

**Comment by hostep** [comment]
@nathanjosiah: ah that's good to know! Have those commits already been pushed to github? If yes, can you point to them? So we can patch our projects. My search for "AC-13493" on github results in nothing at the moment.

**Comment by hostep** [comment]
Here's the merge commit for the people interested, which landed in the `2.4-develop` branch today: https://github.com/magento/magento2/commit/87d012e50c28f448ef1b33ddf50e85d9b3d0a72e
